tasks/ast_tools/src/to_code.rs

```diff
@@ -0,0 +1,189 @@
+use std::mem;
+
+use proc_macro2::{Delimiter, Group, Ident, Literal, Punct, Spacing, TokenStream, TokenTree};
+use quote::quote;
+use syn::Lit;
+
+/// Similar to `quote!` macro, except evaluates to a `TokenStream` of code to *create* the `TokenStream`
+/// which `quote!` evaluates to.
+///
+/// `code!( foo() )` evaluates to a `TokenStream` containing:
+///
+/// ```
+/// TokenStream::from_iter([
+///     TokenTree::Ident(Ident::new("foo", Span::call_site())),
+///     TokenTree::Group(Group::new(Delimiter::Parenthesis, TokenStream::new()))
+/// ].into_iter())
+/// ```
+///
+/// `@{...}` can be used to inject code into output without conversion.
+/// The value in `@{...}` must be a `TokenStream`.
+/// e.g. `code!( foo( @{args} ) )` evaluates to a `TokenStream` containing:
+///
+/// ```
+/// TokenStream::from_iter([
+///     TokenTree::Ident(Ident::new("foo", Span::call_site())),
+///     TokenTree::Group(Group::new(Delimiter::Parenthesis, args))
+/// ].into_iter())
+/// ```
+macro_rules! code {
```

Comment:
This is super smart! I like what I'm seeing👍

Reply:
Thanks. But unfortunately it's not moving the needle on performance.

I suspect (but don't know) that creating `TokenTree`s and `TokenStream`s is inherently expensive due to how proc macros interface with the compiler. There may just be no way to make proc macros cheap.

Or maybe the `#[ast]` macro just isn't that costly in the first place, so optimizing it is never going to pay dividends.

Thanks for taking a look. I'm going to pause on this though for now as am planning to spend this week focused on transformer. Must... not... get... distracted...

By the way, these 2 articles were the inspiration. They do a very good job of explaining how proc macros work internally. I always thought it involved serialization, but it turns out not:
* https://blog.jetbrains.com/rust/2022/03/18/procedural-macros-under-the-hood-part-i/
* https://blog.jetbrains.com/rust/2022/07/07/procedural-macros-under-the-hood-part-ii/

Reply:
> Or maybe the #[ast] macro just isn't that costly in the first place, so optimizing it is never going to pay dividends.

I think this is more likely; No matter how much proc macro overhead costs us, If we do enough work in a macro eventually caching it would be faster. I suspect we have a better chance of getting some juice out of this approach in #4112 However it might be a wild goose chase.

And thanks for the links, They seem really interesting especially the second part(gave them a cursory look), I'll give them a read.